File: MineInAbyss/src/main/java/com/derongan/minecraft/mineinabyss/ascension/effect/effects/ParticleAscensionEffect.java
package com.derongan.minecraft.mineinabyss.ascension.effect.effects;

import org.bukkit.entity.Player;

import java.util.List;

//TODO fix for 1.13
public class ParticleAscensionEffect extends AbstractAscensionEffect {
//    private List<EnumParticle> particles;

    public ParticleAscensionEffect(int offset, int strength, int duration, int iterations, List<Object> particleList) {
        super(offset, strength, duration, iterations);
//        particles = particleList;
    }

    @Override
    public void applyEffect(Player player) {
//        for(EnumParticle p:particles){ addParticlesAroundHead(player, p); }
    }

//    private void addParticlesAroundHead (Player player, EnumParticle particle) {
//        Location loc = player.getEyeLocation();
//
//        float x = (float) loc.getX();
//        float y = (float) loc.getY();
//        float z = (float) loc.getZ();
//
//        PacketPlayOutWorldParticles p = new PacketPlayOutWorldParticles(particle, true, x, y, z, .5f, .5f, .5f, 0f, (int)strength*5);
//
//        ((CraftPlayer) player).getHandle().playerConnection.sendPacket(p);
//    }
}

File: src/main/java/com/derongan/minecraft/mineinabyss/ascension/effect/effects/ParticleAscensionEffect.kt
package com.derongan.minecraft.mineinabyss.ascension.effect.effects

import com.derongan.minecraft.mineinabyss.ascension.effect.AbstractAscensionEffect
import com.mineinabyss.idofront.destructure.component1
import com.mineinabyss.idofront.destructure.component2
import com.mineinabyss.idofront.destructure.component3
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.bukkit.Particle
import org.bukkit.entity.Player

//TODO not sure if anything else needs to be updated with this
@Serializable
@SerialName("particles")
data class ParticleAscensionEffect(
        val strength: Int,
        override val offset: Long,
        override val duration: Int,
        override val iterations: Int,
        private val particles: List<Particle>
) : AbstractAscensionEffect() {
    override fun applyEffect(player: Player) {
        particles.forEach { addParticlesAroundHead(player, it) }
    }

    private fun addParticlesAroundHead(player: Player, particle: Particle) {
        val (x, y, z) = player.eyeLocation
        player.spawnParticle(particle, x, y, z, strength * 5, .5, .5, .5, 0f)
    }
    override fun clone() = copy()
}